Implementing of weather forecast

tfilo opened the pull request.

tfilo commented:
At first it was my idea too, but server side fetching will be better in my opinion. It will not increase number of api request when multiple users wisit site. And since forecast is in hour intervals i will try later to further optimize it by fetching data only once in hour and cache it. Not sure how but i will figure out something i hope.

seehausen commented on skins/neowx-material/skin.conf:
```diff
@@ -276,6 +276,29 @@
+        #    in case you installed weewx with apt: "apt install python3-requests"
+        
+        # Enable forecast yes / no
+        enable_forecast = no
```
```suggestion
        enable = no
```
I would change the parameter to `enable`
we are in the context of forecast so "enable_forecast" is double

seehausen replied:
need changes probably at more places then this ;-)

seehausen commented:
@tfilo 
what do you think about this header?
<img width="1465" height="282" alt="image" src="https://github.com/user-attachments/assets/aee077bc-12b6-4b21-91b0-b017510b2576" />

seehausen commented:
Great, I will do some test today

tfilo commented:
I added following changes:

- merged master into this branch
- added default value for "type" if missconfigures, simple will be used and info message will be logged
- simplified configuration and make it not case sensitive anymore
- updated instructions and search_list_extensions part
- updated DE translation
- added some try/catch into openmeteo.py and added cleaning cache if some of this error occures
- added precipation to simple forecast

I was thinking about adding location into forecast card, but since it generates for same location as $station.location which is already in header, isn't it duplicity ? I would omit it inside card with forecast. I will do some testing of this changes.

tfilo commented:
> @tfilo what do you think about this footer? <img alt="image" width="1465" height="282" src="https://private-user-images.githubusercontent.com/34270572/494173865-aee077bc-12b6-4b21-91b0-b017510b2576.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTg5MDE2NTYsIm5iZiI6MTc1ODkwMTM1NiwicGF0aCI6Ii8zNDI3MDU3Mi80OTQxNzM4NjUtYWVlMDc3YmMtMTJiNi00YjIxLTkxYjAtYjAxNzUxMGIyNTc2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTI2VDE1NDIzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFlZGIwN2U3YmUyMGZmYWY3MjNlMGNhYzdiNGEyZjA1N2NmOWU0ZDlkOGRiMzhjOTZiM2E4OWIwMzIzZDAxMDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.p8onvMhIr4xByteVmXNd2BAA1Y8fvBJl-ZRFcwC5Qb4">
> 
> but I would show the "Weather forecast data provided ..." sentence only when forecast is enabled

Sorry i din't saw your comment, I did it already like this:
<img width="1067" height="121" alt="image" src="https://github.com/user-attachments/assets/64374ee9-4adb-450b-ad13-419b49ac68a3" />

it is shown only when forecast enabled. Is it OK or you prefer your version without <hr /> and as second line of footer insead of first ?

seehausen commented:
> > @tfilo what do you think about this footer? 
> > but I would show the "Weather forecast data provided ..." sentence only when forecast is enabled
> 
> Sorry i din't saw your comment, I did it already like this: 
> 
> it is shown only when forecast enabled. Is it OK or you prefer your version without `<hr />` and as second line of footer insead of first ?

hmm I like my proposal more ;-) it is more compact, and if I read from Top -> Bottom "our" information should be visible before legal text... (and without <hr>

but it is not a blocker

tfilo commented:
Footer fixed as you suggested

seehausen commented on bin/user/openmeteo.py:
```diff
@@ -0,0 +1,248 @@
+#
+# Copyright (c) 2025 Tomáš Filo <tfilosk@gmail.com>
+#
+# Distributed under the terms of the GNU GENERAL PUBLIC LICENSE
+#
```
I could reproduce the error from last test
It is still there, but since you have a try/catch it fixes itselfe with the next run (without need to restart weewx) so it is not a major

to reproduce:
configure forecast with `type = simple`
then later switch to     `type = Advanced`.  (nice it now supports any case :-) )

then you will see
```
Sep 26 19:15:25 weewx weewxd[3975418]: INFO user.historygenerator: historygenerator.py: Generated 12 tables in 3.01 seconds
Sep 26 19:16:02 weewx weewxd[3975418]: ERROR user.openmeteo: Error occured while processing forecast data, cleaning cached data. If this error occure again, please see detailed log below.
Sep 26 19:16:02 weewx weewxd[3975418]: ERROR user.openmeteo: list index out of range
Sep 26 19:19:17 weewx weewxd[3975418]: INFO weewx.cheetahgenerator: Generated 14 files for report NeowxMaterial in 234.62 seconds
Sep 26 19:19:17 weewx weewxd[3975418]: INFO weewx.reportengine: Copied 0 files to /var/www/html/weewx
```
I haven't had a deeper look, but I think you only cache the values you need, and if you switch to "advance" there are some values missing... maybe just cache all, or if you run into the catch case, just recall the api...
but as I said... just a minor thing

tfilo replied:
I will fix it. I have idea how to simply fix this.

seehausen commented on bin/user/openmeteo.py:
```diff
@@ -0,0 +1,248 @@
+        params["hourly"] = ",".join(hourly)
+        params["daily"] = ",".join(daily)
+
+        log.debug("params: %s", params)
```
Is there a way to ignore this "finding" I do not see it critical, and we can dismiss it but maybe there is something like

"@ignorr_scan_warning"

seehausen replied:
or just let us ignore it with false positive or won'f fix
<img width="477" height="578" alt="image" src="https://github.com/user-attachments/assets/ab6f617c-7b00-4aad-a5a0-ff9fbaef3758" />

seehausen replied:
dissmieed

seehausen reviewed (commented):
Looks all good, added one small possible enhancement regarding a proper version for your openmeteo.py

Today I will test the installation (from scratch and update) but then... 
Squash your commits and we will merge...

Can you write a nice release note ;-)

tfilo commented:
I tried fresh instalation in my testing debian 13 instalation today. No problem there. I did try update only while development. When you aprove i will than squash and merge. Not sure about your small change, did you commit it already? I will come up with some release note, no problem there.

BliXem1 commented:
If Dutch translation is needed, let me know. I can help with that.

tfilo commented:
So i did last commit, I squashed all commits into single commit and force pushed it into this branch. Implementation should be done unless you find some bug.

I did this additional changes:

- days variable introduced allowing to set 1 to 7 days of forecast
- updated doc
- migrated from request library to urllib.request, it is more low level library but it is already used with weewx too so it will not require any additional installation from users.
- added up to 3 retries if fetching of forecast fails (it occurs for me from time to time while testing)

seehausen commented:
And for discussion a compact design proposal

<img width="462" height="162" alt="image" src="https://github.com/user-attachments/assets/05877c28-fecf-4f61-805e-15b61b7cc90c" />

tfilo commented:
> Very impressive, I will have a look later
> 
> maybe we do not need translation and use normal "time formats"
> 
> ```
> 
>           [[[[TimeFormats]]]]
> 
>                 # day        = %H:%M
> 
>                 # week       = %H:%M on %A
> 
>                 # month      = %d-%b-%Y %H:%M
> 
> 
> 
>                dayname = %A 
> 
> 
> 
> %A | Full weekday name. | Sunday, Monday,
> 
> ```
> 
> and then use 
> 
> `getText("forecast") + " - " +  date.format($Extras.Formatting.dayname)` 
> 
>    ----> `Daily forecast - Monday`

I did it first with %A but it is in incorrect format. In my language at least there is word for Sunday: Nedela which is base format, but in sentence Forecast for Sunday it us Nedel**u** thats why i introduced translations. I think you have something similar for german language as well.

But with dash it can be ok in base format too.

tfilo commented:
> 
> I must say, it looks good! For me the icon is good enough. Question, can you also change what you want to display with the icons? Like rain, temperature, of nothing only the icons?
> 
>

Well this can be tricky, at first i wanted to implement it in way it will be completly free to include any supported fields from open-meteo but it was too complicated. But for limited number of options it can be possible. Let me think about it. I am away from home for rest of weekend but i will at least think about it how to implement and keep it simple.

seehausen commented:
> I did it first with %A but it is in incorrect format. In my language at least there is word for Sunday: Nedela which is base format, but in sentence Forecast for Sunday it us Nedel**u** thats why i introduced translations. I think you have something similar for german language as well.
> 
> But with dash it can be ok in base format too.

Yes I have the same problem, but it's weewx.
weewx uses the days format and language from the underlaying OS that's why it is in english (at least on my Raspi)
this can be changed be changing default language in your OS... 
I see also english "month" in the history if I switch to german, I could switch to german 
Think could also fixed with setting ENV variable
https://weewx.com/docs/5.1/custom/localization/#environment-variable-LANG

So taking %A would be fine, if it works with setting `export LANG=de_DE.UTF-8`
I can check this

BliXem1 commented:
weectl extension install https://github.com/seehase/neowx-material/archive/refs/heads/129_weather_forecast.zip is everything I can do so I can test it out too? Or is it better to wait first until the release? :D

seehausen commented:
> weectl extension install https://github.com/seehase/neowx-material/archive/refs/heads/129_weather_forecast.zip is everything I can do so I can test it out too? Or is it better to wait first until the release? :D

Thats all... restart weewx after installation

You need to adjust some parameters, thats it

seehausen commented on skins/neowx-material/skin.conf:
```diff
@@ -380,7 +398,7 @@
 #
 [CheetahGenerator]
-    search_list_extensions = user.historygenerator.MyXSearch
+    search_list_extensions = user.historygenerator.MyXSearch, user.openmeteo.Forecast
 
```
It's now enabled by default
please 
```suggestion
    search_list_extensions = user.historygenerator.MyXSearch
    # to enable forecast uncomment next line and comment line above....
    # search_list_extensions = user.historygenerator.MyXSearch, user.openmeteo.Forecast
```

seehausen commented on skins/neowx-material/lang/en.conf:
```diff
@@ -87,3 +87,48 @@
     yesterday            = Yesterday
+
+    [[Forecast]]
+        forecast_1       = Forecast for Monday
```
maybe only `"forecast = Forcast for"` is needed

BliXem1 commented:
> > weectl extension install https://github.com/seehase/neowx-material/archive/refs/heads/129_weather_forecast.zip is everything I can do so I can test it out too? Or is it better to wait first until the release? :D
> 
> Thats all... restart weewx after installation
> 
> You need to adjust some parameters, thats it

Maybe stupid question, and not the right thread, but uhm, when I install this, is it not overwritten any files which already excist?

seehausen commented:
or how about

<img width="471" height="177" alt="image" src="https://github.com/user-attachments/assets/b3d194e9-61ac-4b8f-8609-48ff7102238e" />

seehausen commented:
> weectl extension install https://github.com/seehase/neowx-material/archive/refs/heads/129_weather_forecast.zip

Checked it on mobile... then with logo will be to small... but we can use responsive design, 
`if width is more then xx px show Icon` else not

I think i could work out

seehausen commented:
@tfilo 
Really like you proposal, looks perfect.
and the idea with setup of variables to show  👍

Regarding the 3 column thing with icon:
What about let them define 3 columns, and if size is to small, we just hide the icons if it dose not fit anymore (responsive on breakpoints)... so they can see it, if it fits, on smaller display it is hidden, that would be (my) perfekt solution

Regarding breaking change:
I thought about enable the new feature by default, so in our skin.conf we have everything enabled
in `[[Appearance]]` section and `CheetahGenerator`

If you update via `weectl extension install` everything works out of the box
for those who just copy the files we need to provide a small "migration" script e.g. copy openmeteo.py to bin/user
and change xxx in skin.conf
We can also give this a new major version 2.0 if you like

stalkerGH commented:
I found this option on open-meteo.com: https://open-meteo.com/en/docs

You have to go to _Weather models_ tab, select one of them (checkbox) and scroll down to _API Response_ (look for API URL), just above _Data sources_

Example: [ECMWF HiRes](https://api.open-meteo.com/v1/forecast?latitude=52.52&longitude=13.41&hourly=temperature_2m&models=ecmwf_ifs) or [DWD ICON](https://api.open-meteo.com/v1/forecast?latitude=52.52&longitude=13.41&hourly=temperature_2m&models=icon_seamless)
